File: macos/Sources/WGFileSender/Core/SendClient.swift
import Foundation
import CryptoKit

enum WGFSError: LocalizedError {
    case http(Int)
    case declined
    case badResponse
    case unreachable(String)
    case incomplete        // server closed before the whole file arrived — resumable
    case checksum          // fully received but hash mismatched — corrupt

    var errorDescription: String? {
        switch self {
        case .http(let c): return "Server returned \(c)"
        case .declined: return "The other device declined"
        case .badResponse: return "Unexpected response"
        case .unreachable(let s): return "Can't reach device: \(s)"
        case .incomplete: return "Connection ended early"
        case .checksum: return "Checksum mismatch"
        }
    }
}

/// Outbound side: probes, pairing, sending a manifest (`/offer`), and — as the receiver —
/// pulling file bytes (`/pull`) with resume + hash verification.
final class SendClient {
    private let config: SharedConfig
    init(config: SharedConfig) { self.config = config }

    func ping(address: String) async throws -> PingResponse {
        let (data, resp) = try await dataTask(url: "http://\(address)/ping")
        try check(resp, expect: 200)
        guard let r = try? JSONDecoder().decode(PingResponse.self, from: data) else {
            throw WGFSError.badResponse
        }
        return r
    }

    func requestPair(address: String, sessionId: String, pin: String) async throws -> PairAcceptResponse {
        let id = config.identity
        let body = PairRequestBody(deviceId: id.deviceId, deviceName: id.deviceName,
                                   platform: id.platform, sessionId: sessionId, pin: pin,
                                   port: config.settings.port)
        var req = URLRequest(url: URL(string: "http://\(address)/pair/request")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(body)
        req.timeoutInterval = 90
        let (data, resp) = try await URLSession.shared.data(for: req)
        if let http = resp as? HTTPURLResponse, http.statusCode == 403 { throw WGFSError.declined }
        try check(resp, expect: 200)
        guard let r = try? JSONDecoder().decode(PairAcceptResponse.self, from: data) else {
            throw WGFSError.badResponse
        }
        return r
    }

    func confirmPair(address: String, tokenOut: String, ourTokenForPeer: String) async throws {
        let id = config.identity
        let body = PairConfirmBody(deviceId: id.deviceId, token: ourTokenForPeer)
        var req = URLRequest(url: URL(string: "http://\(address)/pair/confirm")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(tokenOut)", forHTTPHeaderField: "Authorization")
        req.httpBody = try JSONEncoder().encode(body)
        let (_, resp) = try await URLSession.shared.data(for: req)
        try check(resp, expect: 204)
    }

    // MARK: sender — announce a batch manifest

    func offer(to peer: PeerDevice, batchId: String, files: [OfferFile]) async throws {
        let id = config.identity
        var req = URLRequest(url: URL(string: "http://\(peer.peerAddress)/offer")!)
        req.httpMethod = "POST"
        req.setValue("Bearer \(peer.tokenOut)", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(id.deviceId, forHTTPHeaderField: "X-WGFS-Device-Id")
        req.setValue(id.deviceName, forHTTPHeaderField: "X-WGFS-Device-Name")
        req.httpBody = try JSONEncoder().encode(OfferBody(batchId: batchId, files: files))
        req.timeoutInterval = 30
        let (_, resp) = try await URLSession.shared.data(for: req)
        try check(resp, expect: 200)
    }

    // MARK: receiver — pull a file's bytes, resuming and verifying

    /// Pulls (or resumes) one file into `downloadRoot/<senderFolder>/<fileName>` and returns
    /// the saved path. Throws `.incomplete` (keep .part, retry) or `.checksum` (corrupt).
    func pullFile(from peer: PeerDevice, batchId: String, fileId: String, fileName: String,
                  expectedSize: Int64, expectedHash: String, senderFolder: String,
                  downloadRoot: String, progress: @escaping (Int64, Int64) -> Void) async throws -> String {
        let fm = FileManager.default
        let partURL = partURL(fileId)
        try? fm.createDirectory(at: partURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        let start = partFileSize(fileId)   // resume from whatever we already hold

        if start >= expectedSize, start > 0 {
            // Already have all the bytes; just verify + finalize.
            return try finalize(fileId: fileId, fileName: fileName, expectedSize: expectedSize,
                                expectedHash: expectedHash, senderFolder: senderFolder, downloadRoot: downloadRoot)
        }

        var req = URLRequest(url: URL(string: "http://\(peer.peerAddress)/pull?batch=\(batchId)&file=\(fileId)")!)
        req.setValue("Bearer \(peer.tokenOut)", forHTTPHeaderField: "Authorization")
        if start > 0 { req.setValue("bytes=\(start)-", forHTTPHeaderField: "Range") }
        req.timeoutInterval = 60   // stall timeout; resets while bytes flow

        let delegate = PullProgressDelegate(baseline: start, total: expectedSize, onProgress: progress)
        let (tempURL, resp) = try await URLSession.shared.download(for: req, delegate: delegate)
        guard let http = resp as? HTTPURLResponse, http.statusCode == 200 || http.statusCode == 206 else {
            throw WGFSError.http((resp as? HTTPURLResponse)?.statusCode ?? 0)
        }

        // Append the freshly downloaded range onto the .part.
        if start > 0 {
            guard let wh = try? FileHandle(forWritingTo: partURL) else { throw WGFSError.badResponse }
            try? wh.seekToEnd()
            if let th = try? FileHandle(forReadingFrom: tempURL) {
                while case let d = th.readData(ofLength: 1 << 20), !d.isEmpty { try? wh.write(contentsOf: d) }
                try? th.close()
            }
            try? wh.close()
            try? fm.removeItem(at: tempURL)
        } else {
            try? fm.removeItem(at: partURL)
            try fm.moveItem(at: tempURL, to: partURL)
        }

        guard partFileSize(fileId) >= expectedSize else { throw WGFSError.incomplete }
        return try finalize(fileId: fileId, fileName: fileName, expectedSize: expectedSize,
                            expectedHash: expectedHash, senderFolder: senderFolder, downloadRoot: downloadRoot)
    }

    /// Verifies the completed `.part` against the expected hash and moves it into place.
    private func finalize(fileId: String, fileName: String, expectedSize: Int64,
                          expectedHash: String, senderFolder: String, downloadRoot: String) throws -> String {
        let fm = FileManager.default
        let partURL = partURL(fileId)
        let digest = try sha256(partURL)
        guard digest == expectedHash.lowercased() else {
            try? fm.removeItem(at: partURL)   // corrupt — start over next round
            throw WGFSError.checksum
        }
        let folder = URL(fileURLWithPath: downloadRoot)
            .appendingPathComponent(safeComponent(senderFolder), isDirectory: true)
        try? fm.createDirectory(at: folder, withIntermediateDirectories: true)
        let finalURL = uniqueURL(in: folder, fileName: fileName)
        try fm.moveItem(at: partURL, to: finalURL)
        return finalURL.path
    }

    // MARK: helpers

    func fileSize(_ url: URL) -> Int64 {
        let attrs = try? FileManager.default.attributesOfItem(atPath: url.path)
        return (attrs?[.size] as? Int64) ?? 0
    }

    func sha256(_ url: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }
        var hasher = SHA256()
        while true {
            let chunk = handle.readData(ofLength: 1 << 20)
            if chunk.isEmpty { break }
            hasher.update(data: chunk)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }

    private func partURL(_ fileId: String) -> URL {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("WGFileSender/parts", isDirectory: true)
            .appendingPathComponent("\(fileId).part")
    }

    private func partFileSize(_ fileId: String) -> Int64 {
        let attrs = try? FileManager.default.attributesOfItem(atPath: partURL(fileId).path)
        return (attrs?[.size] as? Int64) ?? 0
    }

    private func safeComponent(_ name: String) -> String {
        let cleaned = name.replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "..", with: "_")
        return cleaned.isEmpty ? "Unknown" : cleaned
    }

    private func uniqueURL(in folder: URL, fileName: String) -> URL {
        let fm = FileManager.default
        let safe = safeComponent(fileName)
        var candidate = folder.appendingPathComponent(safe)
        guard fm.fileExists(atPath: candidate.path) else { return candidate }
        let ext = (safe as NSString).pathExtension
        let stem = (safe as NSString).deletingPathExtension
        var n = 1
        repeat {
            let name = ext.isEmpty ? "\(stem) (\(n))" : "\(stem) (\(n)).\(ext)"
            candidate = folder.appendingPathComponent(name)
            n += 1
        } while fm.fileExists(atPath: candidate.path)
        return candidate
    }

    private func dataTask(url: String) async throws -> (Data, URLResponse) {
        guard let u = URL(string: url) else { throw WGFSError.unreachable(url) }
        do { return try await URLSession.shared.data(from: u) }
        catch { throw WGFSError.unreachable(error.localizedDescription) }
    }

    private func check(_ resp: URLResponse, expect: Int) throws {
        guard let http = resp as? HTTPURLResponse else { throw WGFSError.badResponse }
        guard http.statusCode == expect else { throw WGFSError.http(http.statusCode) }
    }
}

/// Reports pull progress (bytes already held + bytes written this request).
final class PullProgressDelegate: NSObject, URLSessionDownloadDelegate {
    private let baseline: Int64
    private let total: Int64
    private let onProgress: (Int64, Int64) -> Void
    init(baseline: Int64, total: Int64, onProgress: @escaping (Int64, Int64) -> Void) {
        self.baseline = baseline
        self.total = total
        self.onProgress = onProgress
    }
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask,
                    didFinishDownloadingTo location: URL) {}
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask,
                    didWriteData bytesWritten: Int64, totalBytesWritten: Int64,
                    totalBytesExpectedToWrite: Int64) {
        onProgress(baseline + totalBytesWritten, total)
    }
}

File: android/app/src/main/java/com/jaewonbaek/wgfilesender/net/SendClient.kt
package com.jaewonbaek.wgfilesender.net

import android.content.Context
import android.net.Uri
import android.webkit.MimeTypeMap
import androidx.documentfile.provider.DocumentFile
import com.jaewonbaek.wgfilesender.data.SharedConfig
import com.jaewonbaek.wgfilesender.model.OfferBody
import com.jaewonbaek.wgfilesender.model.OfferFile
import com.jaewonbaek.wgfilesender.model.PairAcceptResponse
import com.jaewonbaek.wgfilesender.model.PairConfirmBody
import com.jaewonbaek.wgfilesender.model.PairRequestBody
import com.jaewonbaek.wgfilesender.model.PeerDevice
import com.jaewonbaek.wgfilesender.model.PingResponse
import io.ktor.client.HttpClient
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.HttpTimeout
import io.ktor.client.plugins.timeout
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.prepareGet
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsChannel
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.contentType
import io.ktor.utils.io.readAvailable
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.security.MessageDigest

class PullIncompleteException : Exception("Connection ended early")
class ChecksumException : Exception("Checksum mismatch")

/** Outbound side: probes, pairing, announcing a manifest (/offer), and — as the receiver —
 *  pulling file bytes (/pull) with resume + hash verification. */
class SendClient(private val context: Context, private val config: SharedConfig) {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val client = HttpClient(CIO) { install(HttpTimeout) }

    suspend fun ping(address: String): PingResponse {
        val resp = client.get("http://$address/ping")
        check(resp.status.value == 200) { "ping ${resp.status.value}" }
        return json.decodeFromString(resp.bodyAsText())
    }

    suspend fun requestPair(address: String, sessionId: String, pin: String): PairAcceptResponse {
        val id = config.identity
        val body = PairRequestBody(id.deviceId, id.deviceName, id.platform, sessionId, pin, config.settings.port)
        val resp = client.post("http://$address/pair/request") {
            contentType(ContentType.Application.Json)
            setBody(json.encodeToString(body))
            timeout { requestTimeoutMillis = 90_000 }
        }
        if (resp.status.value == 403) throw PairDeclinedException()
        check(resp.status.value == 200) { "pair ${resp.status.value}" }
        return json.decodeFromString(resp.bodyAsText())
    }

    suspend fun confirmPair(address: String, tokenOut: String, ourTokenForPeer: String) {
        val body = PairConfirmBody(config.identity.deviceId, ourTokenForPeer)
        val resp = client.post("http://$address/pair/confirm") {
            contentType(ContentType.Application.Json)
            header(HttpHeaders.Authorization, "Bearer $tokenOut")
            setBody(json.encodeToString(body))
        }
        check(resp.status.value == 204) { "confirm ${resp.status.value}" }
    }

    // MARK: sender — announce a batch manifest

    suspend fun offer(peer: PeerDevice, batchId: String, files: List<OfferFile>) {
        val id = config.identity
        val resp = client.post("http://${peer.peerAddress}/offer") {
            header(HttpHeaders.Authorization, "Bearer ${peer.tokenOut}")
            header("X-WGFS-Device-Id", id.deviceId)
            header("X-WGFS-Device-Name", id.deviceName)
            contentType(ContentType.Application.Json)
            setBody(json.encodeToString(OfferBody(batchId, files)))
            timeout { requestTimeoutMillis = 30_000 }
        }
        check(resp.status.value == 200) { "offer ${resp.status.value}" }
    }

    /** Hashes and measures a content-uri in one pass (for building the manifest). */
    fun hashAndSize(uri: Uri): Pair<String, Long> {
        val md = MessageDigest.getInstance("SHA-256")
        var total = 0L
        context.contentResolver.openInputStream(uri)?.use { input ->
            val buf = ByteArray(1 shl 16)
            while (true) {
                val n = input.read(buf)
                if (n < 0) break
                md.update(buf, 0, n)
                total += n
            }
        }
        return md.digest().joinToString("") { "%02x".format(it) } to total
    }

    // MARK: receiver — pull a file's bytes, resuming and verifying

    /** Pulls (or resumes) one file into the SAF tree and returns the saved uri. Throws
     *  [PullIncompleteException] (keep .part, retry) or [ChecksumException] (corrupt). */
    suspend fun pullFile(
        peer: PeerDevice, batchId: String, fileId: String, fileName: String,
        expectedSize: Long, expectedHash: String, senderFolder: String,
        onProgress: (Long) -> Unit
    ): String {
        val part = File(context.cacheDir, "$fileId.part")
        val start = if (part.exists()) part.length() else 0L

        if (start < expectedSize || start == 0L) {
            client.prepareGet("http://${peer.peerAddress}/pull?batch=$batchId&file=$fileId") {
                if (start > 0) header(HttpHeaders.Range, "bytes=$start-")
            }.execute { resp ->
                if (resp.status.value !in 200..299) throw IOException("pull ${resp.status.value}")
                val channel = resp.bodyAsChannel()
                withContext(Dispatchers.IO) {
                    FileOutputStream(part, /* append = */ start > 0L).use { out ->
                        val buf = ByteArray(1 shl 16)
                        var received = start
                        while (true) {
                            val n = channel.readAvailable(buf, 0, buf.size)
                            if (n == -1) break
                            if (n > 0) {
                                out.write(buf, 0, n)
                                received += n
                                onProgress(received)
                            }
                        }
                    }
                }
            }
        }

        if (part.length() < expectedSize) throw PullIncompleteException()
        val digest = sha256(part)
        if (digest != expectedHash.lowercase()) {
            part.delete()   // corrupt — start over next round
            throw ChecksumException()
        }
        val saved = saveToTree(senderFolder, fileName, part)
            ?: throw IOException("no download folder")
        part.delete()
        return saved
    }

    private fun saveToTree(folderName: String, fileName: String, source: File): String? {
        val treeUriStr = config.settings.downloadTreeUri ?: return null
        val tree = DocumentFile.fromTreeUri(context, Uri.parse(treeUriStr)) ?: return null
        val safeFolder = folderName.replace('/', '_').ifBlank { "Unknown" }
        val folder = tree.findFile(safeFolder)?.takeIf { it.isDirectory }
            ?: tree.createDirectory(safeFolder) ?: return null
        val safeName = fileName.replace('/', '_').replace("..", "_").ifBlank { "file" }
        val ext = safeName.substringAfterLast('.', "")
        val mime = MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext.lowercase())
            ?: "application/octet-stream"
        val target = folder.createFile(mime, safeName) ?: return null
        val out = context.contentResolver.openOutputStream(target.uri) ?: return null
        out.use { o -> source.inputStream().use { it.copyTo(o) } }
        return target.uri.toString()
    }

    private fun sha256(file: File): String {
        val md = MessageDigest.getInstance("SHA-256")
        file.inputStream().use { input ->
            val buf = ByteArray(1 shl 16)
            while (true) {
                val n = input.read(buf)
                if (n < 0) break
                md.update(buf, 0, n)
            }
        }
        return md.digest().joinToString("") { "%02x".format(it) }
    }
}
